src/style/fog.js

```diff
@@ -29,16 +34,77 @@ const TRANSITION_SUFFIX = '-transition';
 export const FOG_PITCH_START = 55;
 export const FOG_PITCH_END = 65;
 
+export class FogSampler {
+    properties: ?PossiblyEvaluated<Props>;
+
+    // As defined in _prelude_fog.fragment.glsl#fog_opacity
+    getFogOpacity(depth: number, pitch: number): number {
+        if (!this.properties) { return 0.0; }
+
+        const props = this.properties;
+        const range = props.get('range');
+        const fogOpacity = props.get('opacity') * smoothstep(FOG_PITCH_START, FOG_PITCH_END, pitch);
+        const start = range[0], end = range[1];
+
+        // The fog is not physically accurate, so we seek an expression which satisfies a
+        // couple basic constraints:
+        //   - opacity should be 0 at the near limit
+        //   - opacity should be 1 at the far limit
+        //   - the onset should have smooth derivatives to avoid a sharp band
+        // To this end, we use an (1 - e^x)^n, where n is set to 3 to ensure the
+        // function is C2 continuous at the onset. The fog is about 99% opaque at
+        // the far limit, so we simply scale it and clip to achieve 100% opacity.
+        // https://www.desmos.com/calculator/3taufutxid
+        const decay = 5.5;
+        let falloff = Math.max(0.0, 1.0 - Math.exp(-decay * (depth - start) / (end - start)));
+
+        // Cube without pow()
+        falloff *= falloff * falloff;
+
+        // Scale and clip to 1 at the far limit
+        falloff = Math.min(1.0, 1.00747 * falloff);
+
+        return falloff * fogOpacity;
+    }
+
+    getOpacityAtTileCoord(x: number, y: number, z: number, tileId: UnwrappedTileID, transform: Transform): number {
+        const mat = transform.calculateCameraMatrix(tileId);
+        const pos = [x, y, z];
+        vec3.transformMat4(pos, pos, mat);
+        const depth = vec3.length(pos);
+
+        return this.getFogOpacity(depth, transform.pitch);
+    }
+
+    getFogOpacityAtLatLng(lngLat: LngLat, transform: Transform): number {
+        const meters = MercatorCoordinate.fromLngLat(lngLat);
+        const elevation = transform.elevation ? transform.elevation.getAtPoint(meters) : 0;
+        const cameraPos = transform._camera.position;
+        const pos = [meters.x - cameraPos[0], meters.y - cameraPos[1], elevation - cameraPos[2]];
+
+        // Transform to pixel coordinate
+        pos[0] *= transform.cameraWorldSize;
+        pos[1] *= transform.cameraWorldSize;
+        pos[2] *= transform.cameraPixelsPerMeter;
```

Comment:
I think some of this logic can be moved into `Transform#_calcMatrices()`theres a `mercatorMatrix` there already for mercator -> screen projection. We can add `mercatorFogMatrix` for mercator -> fog depth projection and just multiply that here. and pass that matrix onto custom layers.